### ISSUE TITLE
New Models API

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#523]
As you noticed, we have some issues with the `MultiHeadPlugin`. The source of this issue is the dynamic change of the head using the labels. Right now, this is done by the plugin. This means that if you call `model(x)` without calling the plugin's callback you get the wrong output because the model does not change the head.

As a solution, I would like to make a custom `DynamicModule`. This is a pytorch `nn.Module` with two additional functionalities.
- a `model_adaptation` phase to add new heads, units and make other architectural changes.
- a `forward` methods that takes `task_labels` as an additional argument.

This approach makes it easy to define dynamic architectures as common in continual learning, such as multi-head or architectural methods that require incremental expansion/freezing of units (e.g. PNN).

There is an important caveat here. Since dynamic modules take an additional argument we can't use standard pytorch modules and substitute the classifier anymore. This is the current API:
```
    model = SimpleMLP(input_size=32*32*3, num_classes=10)
    strategy = Naive(model=model, optimizer=optimizer, criterion=criterion, plugins=[MultiHeadPlugin])
```

This is what will happen with the proposed change:
```
    class MTModel:
        def __init__(feature_extractor):
             self.feature_extractor = feature_extractor
             self.classifier = MultiHeadClassifier(...)

        def forward(x, task_labels):
            x = self.feature_extractor(x)
            return self.classifier(x, task_labels)

    model = MTModel(feature_extractor)
    strategy = Naive(model=model, optimizer=optimizer, criterion=criterion, plugins=[MultiHeadPlugin])
```

The advantages are that:
- multi-task and dynamic models become standalone.
- you can save the model after training with `torch.save` and use `model(x, task_labels)` without the strategy.
- better separation of code which will probably simplify the `BaseStrategy`.
- easier to understand, less bugs that are hard to discover. If you don't pass the `task_labels` you get an error instead of silent feature like we currently have.

The disadvantages are:
- you can't use a pre-made model as is.
- you need to implement a custom forward. It should be easy and mostly boilerplate code but it may be slightly more than we have right now.

Let me know if I should progress on this.

**Comment by vlomonaco** [comment]
It makes sense, I'd go for it!

**Comment by lrzpellegrini** [comment]
Makes sense to mee, too. I'd still try to support pre-made models as much as possible as integrating and maintaining a huge set of model in Avalanche seems a great effort.

@AntonioCarta can we obtain the same behavior by swapping the last fc layer with a `MultiHeadClassifier` at runtime (and monkey-patching the forward implementation, too)?
